### ISSUE TITLE
postgres error

### ISSUE THREAD

**Issue by ctm** [ctm/mb2-doc#1050]
Investigate why we're getting the following error.
```
2022-07-30T01:20:45.754575067Z ERROR [r2d2] another command is already in progress
```
I don't know when this started happening. I only noticed it this morning, but it's clearly been happening since July 30th and I don't have logs from before then. In all cases, the `r2d2` complaint is preceded by:
```
2022-07-30T01:19:42.517099764Z WARN  [mb2::models] Could not insert NewTsidMessage { message: Object({"Private": Object({"DeferredIdentities": Array([Number(0), Array([String("deadhead")])])})}), table_session_id: TableSessionId(2067838930643332462) }: DatabaseError(ForeignKeyViolation, "insert or update on table \"tsid_messages\" violates foreign key constraint \"tsid_messages_table_session_id_fkey\"")
```
and all our `ForeignKeyViolation` error messages have that `r2d2` message after them.

I know the source of the `ForeignKeyViolation` and it's benign. My guess is the `r2d2` message is also benign and that it's either misreporting the `ForeignKeyViolation` itself or it's reporting a real error on a successive access of the database that is still related to the expected-to-fail operation of logging deferred identities for a non-logged in user.

So, although we're probably not losing any info, the exact cause of this needs to be found and fixed, just to be 100% sure. This reinforces the need handle error messages better, especially once caused by `error!` invocations, including those that come from crates that mb2 uses rather than mb2 itself.

**Comment by ctm** [comment]
An obvious possibility is when [`r2d2` updated `parking_lot`](https://github.com/sfackler/r2d2/blob/master/CHANGELOG.md#0810---2022-06-21):
```
https://github.com/sfackler/r2d2/blob/master/CHANGELOG.md#0810---2022-06-21
```

It should be trivial to recreate the `ForeignKeyViolation` locally with the current (0.8.10) `r2d2` as well as with previous version (0.8.9) and see if that's where the error message crept in.

```
commit b8a9d4523444db359ae2628556cddb98f1339a9e
Author: Clifford T. Matthews <ctm@devctm.com>
Date:   Tue Jun 21 21:51:43 2022 -0600

    [Chore] cargo update

diff --git a/Cargo.lock b/Cargo.lock
index 40e9e8a1..03ed14e7 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -29,7 +29,7 @@ dependencies = [
  "futures-util",
  "log",
  "once_cell",
- "parking_lot 0.12.1",
+ "parking_lot",
  "pin-project-lite",
  "smallvec",
  "tokio",
@@ -243,7 +243,7 @@ dependencies = [
  "serde_urlencoded",
  "smallvec",
  "socket2",
- "time 0.3.10",
+ "time 0.3.11",
  "url",
 ]

... 
 
-[[package]]
-name = "parking_lot"
-version = "0.11.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7d17b78036a60663b797adeaee46f5c9dfebb86948d1255007a1d6be0271ff99"
-dependencies = [
- "instant",
- "lock_api",
- "parking_lot_core 0.8.5",
-]
-
 [[package]]
 name = "parking_lot"
 version = "0.12.1"
@@ -1802,21 +1782,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "3742b2c103b9f06bc9fff0a37ff4912935851bee6d36f3c02bcc755bcfec228f"
 dependencies = [
  "lock_api",
- "parking_lot_core 0.9.3",
-]
-
-[[package]]
-name = "parking_lot_core"
-version = "0.8.5"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d76e8e1493bcac0d2766c42737f34458f1c8c50c0d23bcb24ea953affb273216"
-dependencies = [
- "cfg-if",
- "instant",
- "libc",
- "redox_syscall",
- "smallvec",
- "winapi",
+ "parking_lot_core",
 ]
 
 [[package]]
@@ -2009,12 +1975,12 @@ dependencies = [
 
 [[package]]
 name = "r2d2"
-version = "0.8.9"
+version = "0.8.10"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "545c5bc2b880973c9c10e4067418407a0ccaa3091781d1671d46eb35107cb26f"
+checksum = "51de85fb3fb6524929c8a2eb85e6b6d363de4e8c48f9e2c2eac4944abc181c93"
 dependencies = [
  "log",
- "parking_lot 0.11.2",
+ "parking_lot",
  "scheduled-thread-pool",
 ]
 ...
```

**Comment by ctm** [comment]
Unfortunately, when run locally (either natively under macos or under docker), `r2d2` doesn't complain, at least not if I simply make a bogus call to `log_tsid_message` using a non-db `TableSessionId` in `NickMapper::default`. I get the expected `violates foreign key constraint` error message, but nothing from `r2d2`.

**Comment by ctm** [comment]
FWIW, that same hack deployed to production also doesn't trigger the `r2d2` error, so perhaps doing it in `NickMapper::default` isn't sufficient, or perhaps more than a single call to `log_tsid_message` is needed.

**Comment by ctm** [comment]
I've tried to reproduce this locally by signing up three players and having one not be present when the game starts. I then try observing one of the players or the table, but I don't get a `DeferredIdentities` failure. I'll have to go through the code and see if I can figure a way that `DeferredIdentities` gets logged from an account that appears to be real but is really anonymous. It may require a player busting from the tournament and then refreshing the page.

**Comment by ctm** [comment]
I tried starting an NLHE tournament with three players, putting them all-in, then closing the lobby for the player who bust, then refreshing the table page for the bust player. That didn't do it. It just brought up the tourney summary page showing that the bust player had indeed bust.

**Comment by ctm** [comment]
My current guess is that this requires player motion of a non-logged in player to trigger. I can probably simulate this with some extra (conditionally compiled) code. If an anonymous observer is watching table A when it is combined with table B and there's a player who has never been logged in since the anonymous observer started watching, _I think_ an `Identities` request is sent for the nickname of that player who has never logged in.

**Comment by ctm** [comment]
There's probably an easier way to fix this. I should just be able to instrument `Handler<TableSessionRequest>::handle` and locally find out where `is_anon` is incorrectly being set to false. That only causes us trouble when we call `tsid_send`, which we only do to resolve an `Identify` request and that's rare and hard to set up, _but_ just figuring out when `is_anon` is incorrectly set can be done for any call to `handle`, and my guess is that's simply a matter of something like an anonymous observer of a table (as opposed to an anonymous observer of a player).

**Comment by ctm** [comment]
I spent a bunch of time trying to reproduce it locally and I couldn't. However, since `Identify` to a non-anon observer is so rare, I can simply add code that tries to look up the `TableSessionId` in the database when `is_anon` is false and if it's not in the database, log a bunch of info to help track things down and then tell `tsid_send` that it is anonymous after all.

I should have done that in time for this evening's game, but I got sidetracked.

**Comment by ctm** [comment]
I've deployed code that should prevent the foreign key constraint violation from getting hit. It should also provide enough information to figure out why `Handler<TableSessionRequest>::handle` sometimes thinks an anonymous connection is non anonymous. 

Since we never get the `r2d2` error without the foreign key constraint violation, we shouldn't see it again, however, knowing what's going on with `Handler<TableSessionRequest>::handle` _might_ be provide enough information to artificially reproduce the `r2d2` error locally. As such, I may open another issue when that happens.

**Comment by ctm** [comment]
I think there's a good chance that I'll be able to reproduce the `r2d2` error after all due to [what I found](https://github.com/ctm/mb2-doc/issues/1012#issuecomment-1221295142) in #1012.

**Comment by ctm** [comment]
I can't reproduce this locally, but can reproduce it on devctm.com. I didn't think to try reproducing it locally with a production build, but I'll try that after I finish my next experiment, which is temporarily deploying to devctm.com the version of mb2 before the `r2d2` upgrade. Once I know if that version still exhibits the `r2d2` issue, I'll know how to proceed.

**Comment by ctm** [comment]
Turns out I upgraded `r2d2` before I put in the code that has the foreign key violation. So, I'm now temporarily putting up the first version with the foreign key violation to see if it has the `r2d2` complaint. My guess is it will and then I'm left with the possibility that the `r2d2` violation has been there since the beginning (late July) and I just didn't notice it, or something else changed.

My guess is something else changed. Perhaps it was the change in underlying debian versions. That's going to be hard to check, because the security files have been removed, so I can't even build using the old Dockerfile anymore. There may be hoops I can jump through to do so, but I don't know them and probably won't even spend time looking for them (since they may not exist).

If I can reproduce this locally with a production build, I'll track it down that way.

**Comment by ctm** [comment]
My prior local test was with a production build; that's baked into the `Dockerfile`. So, reproducing this locally just isn't going to happen.

Commit `50553a4e` with some minor changes to `Dockerfile`, `Cargo.toml` and `Cargo.lock` *does not* show the `r2d2` error, even when deployed to devctm.com. So, git bisect can find the problem, but it will take a fair amount of time due to how long it takes to deploy. Luckily, with cargo-chef, it's deploys faster and one of the minor changes to `Dockerfile` is adding cargo-chef.

So, I'm going start down the git bisect route and try to do other things while the various deploys go. This will be a PITA, but it's basically psychological (i.e., resentment for having to do this) and I just need to suck it up&hellip;

**Comment by ctm** [comment]
Further complication. Apparently the `r2d2` error does not necessarily come directly after the foreign key constraint violation. There were times where I was doing a git bisect where I was able to create the foreign key constraint violation but not get an immediate `r2d2` error that did give me an `r2d2` error as part of the shutdown process. Unfortunately, that means I need to start the bisect all over again. It also means that I still might get a false positive which would, of course, completely invalidate the git bisect.

This is a bit of a horrorshow.

**Comment by ctm** [comment]
Turns out it was the upgrade to diesel from rc0 to rc1:
```
[(no branch, bisect started on master)]% git bisect bad
b7211c965db7d2b572a1fa2e6802713e9c627d50 is the first bad commit
commit b7211c965db7d2b572a1fa2e6802713e9c627d50
Author: Clifford T. Matthews <ctm@devctm.com>
Date:   Fri Jul 22 05:17:09 2022 -0600

    [Chore] cargo update

 Cargo.lock | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)
[(no branch, bisect started on master)]% git show b7211c9
commit b7211c965db7d2b572a1fa2e6802713e9c627d50 (HEAD, refs/bisect/bad)
Author: Clifford T. Matthews <ctm@devctm.com>
Date:   Fri Jul 22 05:17:09 2022 -0600

    [Chore] cargo update

diff --git a/Cargo.lock b/Cargo.lock
index dfae0ffd..1e697e8a 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -881,9 +881,9 @@ dependencies = [
 
 [[package]]
 name = "diesel"
-version = "2.0.0-rc.0"
+version = "2.0.0-rc.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e531d0abf8147036383c88ce3785d9cda6fe1b8b1183021ffa01cf49c5819d24"
+checksum = "876a12f2d98c35d1dfaf74083664be5bb71606b72f4756f3792cdb1ddb192b46"
 dependencies = [
  "bitflags",
  "byteorder",
@@ -908,9 +908,9 @@ dependencies = [
 
 [[package]]
 name = "diesel_derives"
-version = "2.0.0-rc.0"
+version = "2.0.0-rc.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "212ba0cdc611523c37f26c7b3b4aedb4af7cbfe9239c4e7736a8a7c82284d77b"
+checksum = "2497d9cefebedc40492310f3c94493e1d0bd5b11b5e1bfc0a7f4b106012203ea"
 dependencies = [
  "proc-macro-error",
  "proc-macro2 1.0.40",
@@ -920,9 +920,9 @@ dependencies = [
 
 [[package]]
 name = "diesel_migrations"
-version = "2.0.0-rc.0"
+version = "2.0.0-rc.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e2edcb7e9cff76af48e7fa33cc4680acc7df19e58a4a4aaed3e2b67fb83bc06d"
+checksum = "af3e284b00439e362c3f55849a103f38b4c5a9918ae3cf34fca2268b6eb52fef"
 dependencies = [
  "diesel",
  "migrations_internals",
@@ -1725,9 +1725,9 @@ dependencies = [
 
 [[package]]
 name = "migrations_internals"
-version = "2.0.0-rc.0"
+version = "2.0.0-rc.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "61e2226081f9176165a49b006a0729de8cafa749615b349e15c444dbd39e4d9d"
+checksum = "e03913f1a1c044aa53b37c8f34011599bbf98914092a0444081555e9ca35022c"
 dependencies = [
  "serde",
  "toml",
@@ -1735,9 +1735,9 @@ dependencies = [
 
 [[package]]
 name = "migrations_macros"
-version = "2.0.0-rc.0"
+version = "2.0.0-rc.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "37dbadac6f186c134552cd9a0ec14b157cf4f04eacacebd5e6c1bfc6aeb158b3"
+checksum = "9ccfa634200e9de9fe6497f568d54951e933e9e0ddf61058e52e24bda896de48"
 dependencies = [
  "migrations_internals",
  "proc-macro2 1.0.40",
[(no branch, bisect started on master)]% 
```
In retrospect, the diesel upgrade makes sense at least with respect to the timeframe in which the error popped up. All else equal, I wouldn't have guessed it was diesel, but it's nice to know, and I'll submit an issue to diesel's GitHub repository once I can nail some info down. Not being able to reproduce it locally complicates things a bit.

**Comment by ctm** [comment]
As I was researching my bug report for diesel, I realized mb2 is using a Postgres version 12 AWS server (12.8), but I've been using Postgres version 13 for local development. I brought up Postgres version 12 locally (12.12) and was able to reproduce the error.

As such, I've submitted a [half-baked bug report](https://github.com/diesel-rs/diesel/issues/3282) to diesel which I'll try to fully bake (i.e., create a tiny test case that illustrates the problem) "soon". I'm going to fix the foreign key issue (#1012) "now" and then try to upgrade our RDS instance (#1052), because I'd like to have both of them done for this evening's game if possible, since I'm leaving town tomorrow morning for a couple of days and I won't have very much internet access on Mt. Taylor.
